File: shared/saki-ir/go/sakir/geom.go
package sakir

import (
	"math"
	"sort"

	annotationirv1 "github.com/saki-ai/saki/shared/saki-ir/go/gen/annotationirv1"
)

const screenSortEps = 1e-6

// Point 表示二维坐标点。
type Point struct {
	// X 是点的横坐标。
	X float32
	// Y 是点的纵坐标。
	Y float32
}

// RectTLToCenter 将 Rect(TL) 转换为中心语义 (cx, cy, width, height)。
//
// Spec: docs/IR_SPEC.md#4-rect-semantics
func RectTLToCenter(rect *annotationirv1.RectGeometry) (cx, cy, width, height float32) {
	if rect == nil {
		return 0, 0, 0, 0
	}
	cx = rect.GetX() + rect.GetWidth()/2
	cy = rect.GetY() + rect.GetHeight()/2
	return cx, cy, rect.GetWidth(), rect.GetHeight()
}

// RectCenterToTL 将中心语义转换为 Rect(TL) 的 (x, y, width, height)。
//
// Spec: docs/IR_SPEC.md#4-rect-semantics
func RectCenterToTL(cx, cy, width, height float32) (x, y, w, h float32) {
	x = cx - width/2
	y = cy - height/2
	return x, y, width, height
}

// RectToVerticesScreen 返回屏幕顺序顶点：TL, TR, BR, BL。
//
// Spec: docs/IR_SPEC.md#7-vertices-and-aabb
func RectToVerticesScreen(rect *annotationirv1.RectGeometry) [4]Point {
	if rect == nil {
		return [4]Point{}
	}
	x := rect.GetX()
	y := rect.GetY()
	w := rect.GetWidth()
	h := rect.GetHeight()
	return [4]Point{
		{X: x, Y: y},
		{X: x + w, Y: y},
		{X: x + w, Y: y + h},
		{X: x, Y: y + h},
	}
}

// RectToVertices 是兼容别名，等价 RectToVerticesScreen。
func RectToVertices(rect *annotationirv1.RectGeometry) [4]Point {
	return RectToVerticesScreen(rect)
}

// ObbToVerticesLocal 返回 OBB 局部角点顺序：TL, TR, BR, BL。
//
// 注意：这是局部角点顺序，不是屏幕排序。
//
// Spec: docs/IR_SPEC.md#5-obb-semantics
// Spec: docs/IR_SPEC.md#7-vertices-and-aabb
func ObbToVerticesLocal(obb *annotationirv1.ObbGeometry) [4]Point {
	if obb == nil {
		return [4]Point{}
	}

	cx := float64(obb.GetCx())
	cy := float64(obb.GetCy())
	hw := float64(obb.GetWidth()) / 2.0
	hh := float64(obb.GetHeight()) / 2.0
	theta := float64(obb.GetAngleDegCcw()) * math.Pi / 180.0
	cosT := math.Cos(theta)
	sinT := math.Sin(theta)

	corners := [4][2]float64{
		{-hw, -hh},
		{hw, -hh},
		{hw, hh},
		{-hw, hh},
	}

	var out [4]Point
	for i, c := range corners {
		dx := c[0]
		dy := c[1]
		rx := dx*cosT - dy*sinT
		ry := dx*sinT + dy*cosT
		out[i] = Point{X: float32(cx + rx), Y: float32(cy + ry)}
	}
	return out
}

// ObbToVerticesScreen 返回屏幕排序顶点：TL, TR, BR, BL。
//
// Spec: docs/IR_SPEC.md#7-vertices-and-aabb
func ObbToVerticesScreen(obb *annotationirv1.ObbGeometry) [4]Point {
	local := ObbToVerticesLocal(obb)
	return sortVerticesScreen(local, screenSortEps)
}

// ObbToVertices 是兼容别名，等价 ObbToVerticesLocal。
func ObbToVertices(obb *annotationirv1.ObbGeometry) [4]Point {
	return ObbToVerticesLocal(obb)
}

// VerticesToAABB 由顶点集合计算 AABB (x, y, w, h)。
func VerticesToAABB(vertices [4]Point) (x, y, w, h float32) {
	minX := float64(vertices[0].X)
	minY := float64(vertices[0].Y)
	maxX := minX
	maxY := minY
	for i := 1; i < len(vertices); i++ {
		vx := float64(vertices[i].X)
		vy := float64(vertices[i].Y)
		minX = math.Min(minX, vx)
		minY = math.Min(minY, vy)
		maxX = math.Max(maxX, vx)
		maxY = math.Max(maxY, vy)
	}
	return float32(minX), float32(minY), float32(maxX - minX), float32(maxY - minY)
}

func sortVerticesScreen(vertices [4]Point, eps float64) [4]Point {
	type withKey struct {
		p  Point
		qx int64
		qy int64
	}

	quant := func(v float64) int64 {
		if eps <= 0 {
			return int64(math.Round(v * 1e6))
		}
		return int64(math.Round(v / eps))
	}

	values := make([]withKey, 0, 4)
	for _, p := range vertices {
		values = append(values, withKey{p: p, qx: quant(float64(p.X)), qy: quant(float64(p.Y))})
	}

	sort.Slice(values, func(i, j int) bool {
		if values[i].qy != values[j].qy {
			return values[i].qy < values[j].qy
		}
		if values[i].qx != values[j].qx {
			return values[i].qx < values[j].qx
		}
		if values[i].p.Y != values[j].p.Y {
			return values[i].p.Y < values[j].p.Y
		}
		return values[i].p.X < values[j].p.X
	})

	top := []Point{values[0].p, values[1].p}
	bottom := []Point{values[2].p, values[3].p}
	sort.Slice(top, func(i, j int) bool { return top[i].X < top[j].X })
	sort.Slice(bottom, func(i, j int) bool { return bottom[i].X < bottom[j].X })

	return [4]Point{top[0], top[1], bottom[1], bottom[0]}
}

File: shared/saki-ir/python/src/saki_ir/geom.py
from __future__ import annotations

"""saki-ir 几何计算工具函数。

Spec: docs/IR_SPEC.md#4-rect-semantics
Spec: docs/IR_SPEC.md#5-obb-semantics
Spec: docs/IR_SPEC.md#7-vertices-and-aabb
"""

import math
from typing import TypeAlias

from saki_ir.proto.saki.ir.v1 import annotation_ir_pb2 as annotationirv1

Point: TypeAlias = tuple[float, float]
_SORT_EPS = 1e-6


def rect_tl_to_center(rect: annotationirv1.RectGeometry) -> tuple[float, float, float, float]:
    """将 Rect(TL) 转为 `(cx, cy, w, h)`。"""

    # Spec: docs/IR_SPEC.md#4-rect-semantics
    cx = float(rect.x) + float(rect.width) / 2.0
    cy = float(rect.y) + float(rect.height) / 2.0
    return cx, cy, float(rect.width), float(rect.height)


def rect_center_to_tl(cx: float, cy: float, width: float, height: float) -> tuple[float, float, float, float]:
    """将中心表示转为 Rect(TL) 的 `(x, y, w, h)`。"""

    # Spec: docs/IR_SPEC.md#4-rect-semantics
    x = float(cx) - float(width) / 2.0
    y = float(cy) - float(height) / 2.0
    return x, y, float(width), float(height)


def rect_to_vertices_screen(rect: annotationirv1.RectGeometry) -> list[Point]:
    """返回 Rect 的 4 个顶点，顺序固定为屏幕 `TL, TR, BR, BL`。"""

    # Spec: docs/IR_SPEC.md#7-vertices-and-aabb
    x = float(rect.x)
    y = float(rect.y)
    w = float(rect.width)
    h = float(rect.height)
    return [
        (x, y),
        (x + w, y),
        (x + w, y + h),
        (x, y + h),
    ]


def rect_to_vertices(rect: annotationirv1.RectGeometry) -> list[Point]:
    """兼容别名：等价 `rect_to_vertices_screen`。"""

    return rect_to_vertices_screen(rect)


def obb_to_vertices_local(obb: annotationirv1.ObbGeometry) -> list[Point]:
    """返回 OBB 的 4 个顶点，顺序固定为局部角点顺序 `TL, TR, BR, BL`。

    注意：这里的 `TL/TR/BR/BL` 是 OBB 局部坐标系角点，不是屏幕排序结果。
    """

    # Spec: docs/IR_SPEC.md#5-obb-semantics
    # Spec: docs/IR_SPEC.md#7-vertices-and-aabb
    cx = float(obb.cx)
    cy = float(obb.cy)
    w = float(obb.width)
    h = float(obb.height)
    theta = math.radians(float(obb.angle_deg_ccw))
    cos_t = math.cos(theta)
    sin_t = math.sin(theta)

    half_w = w / 2.0
    half_h = h / 2.0
    corners = [
        (-half_w, -half_h),
        (half_w, -half_h),
        (half_w, half_h),
        (-half_w, half_h),
    ]

    points: list[Point] = []
    for dx, dy in corners:
        # Spec: docs/IR_SPEC.md#5-obb-semantics
        # 屏幕坐标 CCW 旋转（x 右, y 下）
        rx = dx * cos_t - dy * sin_t
        ry = dx * sin_t + dy * cos_t
        points.append((cx + rx, cy + ry))
    return points


def obb_to_vertices_screen(obb: annotationirv1.ObbGeometry, *, eps: float = _SORT_EPS) -> list[Point]:
    """返回 OBB 的屏幕排序角点 `TL, TR, BR, BL`。"""

    # Spec: docs/IR_SPEC.md#7-vertices-and-aabb
    local = obb_to_vertices_local(obb)
    return _sort_vertices_screen(local, eps=eps)


def obb_to_vertices(obb: annotationirv1.ObbGeometry) -> list[Point]:
    """兼容别名：等价 `obb_to_vertices_local`。"""

    return obb_to_vertices_local(obb)


def vertices_to_aabb(vertices: list[Point]) -> tuple[float, float, float, float]:
    """由顶点集合计算 AABB `(x, y, w, h)`。"""

    if not vertices:
        return 0.0, 0.0, 0.0, 0.0
    xs = [float(p[0]) for p in vertices]
    ys = [float(p[1]) for p in vertices]
    x0 = min(xs)
    y0 = min(ys)
    x1 = max(xs)
    y1 = max(ys)
    return x0, y0, x1 - x0, y1 - y0


def _sort_vertices_screen(points: list[Point], *, eps: float) -> list[Point]:
    """按屏幕规则将 4 点排序为 `TL, TR, BR, BL`。"""

    if len(points) != 4:
        raise ValueError("screen 顶点排序要求输入 4 个点")

    scale = 1.0 / eps if eps > 0 else 0.0

    def q(v: float) -> float:
        if scale == 0.0:
            return v
        return round(v * scale)

    by_yx = sorted(points, key=lambda p: (q(float(p[1])), q(float(p[0])), float(p[1]), float(p[0])))
    top = sorted(by_yx[:2], key=lambda p: (q(float(p[0])), float(p[0]), q(float(p[1])), float(p[1])))
    bottom = sorted(by_yx[2:], key=lambda p: (q(float(p[0])), float(p[0]), q(float(p[1])), float(p[1])))
    return [top[0], top[1], bottom[1], bottom[0]]
